Cleanup of operands() accesses

tautschnig opened the pull request:
Although USE_LIST is gone as of r3530, there are a number of cases where iterators are preferable over indexed access to elements.
This is a duplicate of #9 after restoring all links to the SVN.

thk123 commented on src/cpp/cpp_typecheck_expr.cpp:
```diff
@@ -2578,25 +2578,26 @@ void cpp_typecheckt::typecheck_function_call_arguments(
 
-  for(std::size_t i=0; i<parameters.size(); i++)
+  exprt::operandst::iterator arg_it=expr.arguments().begin();
+  for(std::size_t i=0; i<parameters.size(); i++, ++arg_it)
```
Use either prefix or postfix to increment these.

thk123 replied:
Use either prefix or postfix to increment these.

thk123 commented on src/cpp/cpp_typecheck_resolve.cpp:
```diff
@@ -2634,10 +2635,12 @@ void cpp_typecheck_resolvet::resolve_with_arguments(
 {
   // not clear what this is good for
-  for(std::size_t i=0; i<fargs.operands.size(); i++)
+  for(exprt::operandst::const_iterator it=fargs.operands.begin();
```
Can use a ranged for here

thk123 commented on src/solvers/flattening/flatten_byte_operators.cpp:
```diff
@@ -87,7 +87,7 @@ exprt flatten_byte_extract(
       // bits than the original expression will be returned.
       if(width_bytes==1)
-        return op[0];
+        return op.front();
```
Is this preferred to `op0`, if so perhaps there shouldn't even be an `op0`?

thk123 commented on src/util/expr_util.cpp:
```diff
@@ -295,15 +295,18 @@ exprt make_binary(const exprt &expr)
+  exprt previous=operands.front();
   
-  for(std::size_t i=1; i<operands.size(); i++)
+  for(exprt::operandst::const_iterator
```
Range based for.

thk123 commented on src/util/std_code.h:
```diff
@@ -282,7 +282,7 @@ class code_assumet:public codet
   {
     // will change to resize(1) in the future
-    operands().reserve(1);
+    reserve_operands(1);
```
Is the comment above still relevant/correct? Is `reserve_operands` what `resize` ended up being called?

tautschnig commented:
I have now refreshed/rebased this pull request, and have taken into account all the comments as far as applicable.

peterschrammel commented:
@thk123, can you please check whether your comments have been taken into account?

thk123 commented:
@peterschrammel I think these ([1](https://github.com/diffblue/cbmc/pull/23#discussion_r93661140), [3](https://github.com/diffblue/cbmc/pull/23#discussion_r93662655)) comments are still outstanding. Are there reasons for not addressing them @tautschnig?

tautschnig commented:
I have addressed the remaining comments, which induced some further cleanup.

thk123 commented on src/cpp/cpp_typecheck_expr.cpp:
```diff
@@ -2577,28 +2577,30 @@ void cpp_typecheckt::typecheck_function_call_arguments(
         exprt temporary;
-        new_temporary(expr.arguments()[i].source_location(),
-                      parameters[i].type().subtype(),
+        new_temporary(arg_it->source_location(),
```
The parameter should be on the next line. It might be worth rebasing again so you get the lint output from just modified lines.

thk123 commented on src/cpp/cpp_typecheck_fargs.cpp:
```diff
@@ -109,7 +109,8 @@ bool cpp_typecheck_fargst::match(
 
-  for(std::size_t i=0; i<ops.size(); i++)
+  exprt::operandst::iterator it=ops.begin();
+  for(std::size_t i=0; i<ops.size(); i++, ++it)
```
Still mixing post fix and prefix incrementing. Maybe this isn't really a problem (I think it looks weird but it isn't in the coding guidelines) but I thought I'd see this being picked up in other PR reviews.

thk123 commented on src/util/std_code.h:
```diff
@@ -738,13 +737,13 @@ class code_returnt:public codet
   inline bool has_return_value() const
   {
-    return operands().size()==1;
+    if(operands().empty()) return false; // backwards compatibility
```
False on the new line. 
```
if(operands().empty()) 
  return false; // backwards compatibility
```

tautschnig commented:
Thanks a lot for the careful scrutiny! I believe I have addressed the issues, but looking at the cpplint run yields some rather unrelated problems. @thk123 Could you please take a look at the [travis output](https://travis-ci.org/diffblue/cbmc/jobs/193766894)?

thk123 commented:
@tautschnig You seem to be based off a commit that is before the script that just checks the modified lines, if you rebase I imagine they will go away (or at least the only ones left should be legitimate errors). It should also fix the python crash at the end.
